ragexe/shop-services-monorep
# Task 5 - Import to S3

### Task descrition

Task goal - Create a new service, called import-service, with its own serverless-config file at the same level as product service.

_[source](https://github.com/EPAM-JS-Competency-center/cloud-development-course-initial/blob/main/task5-import-to-s3/task.md)_


### What was done

- Initial template creation
- Implement import-products-file lambda mock
- Implement import-products-file lambda function
- Implement unit-tests for import-product-file fn
- Implement lambda import-file-parser
- Implement unit-tests for import-file-parser fn
- Add swagger documentation generation for import-service
- Implement temporary lambda function for a file list check
- Fix product validation after data parsing
- Add valid & invalid `.csv` samples


### Self-review

Reviewers should verify the lambda functions by invoking them through provided URLs.
 
- [x] **1** - File **serverless.yml** contains configuration for **importProductsFile** function
- [x] **3** - The **importProductsFile** lambda function returns a correct response which can be used to upload a file into the **S3** bucket
- [x] **4** - Frontend application is integrated with **importProductsFile** lambda
- [x] **5** - The **importFileParser** lambda function is implemented and **serverless.yml** contains configuration for the lambda

#### Additional (optional) tasks

- [x] **+1** **(for JS only)** - **async/await** is used in lambda functions
- [x] **+1** **(All languages)** - **importProductsFile** lambda is covered by **unit** tests (**(for JS only)** **aws-sdk-mock** can be used to mock S3 methods - https://www.npmjs.com/package/aws-sdk-mock)
- [x] **+1** **(All languages)** - At the end of the **stream** the lambda function should move the file from the **uploaded** folder into the **parsed** folder (move the file means that file should be copied into **parsed** folder, and then deleted from **uploaded** folder)  

_Note: In the current implementation, only files that have been successfully parsed and validated are moved to the `parsed/` directory. You can check the logic with a [\/test](https://llbhmcmrja.execute-api.eu-west-1.amazonaws.com/dev/test) request and the attached files_:  [sample invalid csv](https://github.com/ragexe/shop-services-monorep/blob/8910eb7c24d8160ed7015ca1c4f6f1214fcc8cef/import-service/sample-csv/invalid_products.csv), [sample valid csv](https://github.com/ragexe/shop-services-monorep/blob/8910eb7c24d8160ed7015ca1c4f6f1214fcc8cef/import-service/sample-csv/products0.csv)

Expected result: **8 pts**


### Front-End part

https://github.com/ragexe/shop-angular-cloudfront/pull/4
https://d2dm71pjpz5w21.cloudfront.net/admin/products


### SWAGGER documentation
[product-service_dev](https://app.swaggerhub.com/apis/ragexe/product-service_dev/)
[import-service_dev](https://app.swaggerhub.com/apis/ragexe/import-service_dev/)
